### PR TITLE
refactor: Wallet creation enable single tap of cards

### PR DESCRIPTION
**Pull request by ujjwal-cyph**
Fixes https://app.clickup.com/t/37308523/CHI-2136
Fixes https://app.clickup.com/t/37308523/PRD-1824
Fixes https://app.clickup.com/t/37308523/PRD-1877

### CODE REVIEW

**Comment by vbcypherock** [comment]
Task linked: [CHI-2136 Refactor wallet creation flow with Single tap of cards](https://app.clickup.com/t/37308523/CHI-2136)

**Review by amanCypherock — CHANGES_REQUESTED** [review]

File: src/level_four/tap_cards/controller/write_to_cards_controller.c [hunk 1]
```diff
@@ -66,61 +66,76 @@
...
+        flow_level.level_four = CARD_ONE_WRITE;
+        break;
+
+    case CARD_ONE_WRITE:
+        write_share_to_card(1);
+        break;
+
+    case CARD_ONE_READBACK:
+        readback_share_from_card(0);
         break;
```
[the remark this answers is not in the document]
Comment (ujjwal-cyph) [reply src/level_four/tap_cards/controller/write_to_cards_controller.c:93]: i am refraining to touch the core logic in this PR. can we open a separate ticket?

File: src/level_four/tap_cards/controller/card_action_controllers.c [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+            secret);
+        memzero(wallet_shamir_data.mnemonic_shares, sizeof(wallet_shamir_data.mnemonic_shares));
+    }
+
+    if (status == 1) {
+        // verify wallet id only if secret successfully regenerated
+        mnemonic_clear();
+        const char* mnemo = mnemonic_from_data(wallet.wallet_share_with_mac_and_nonce, wallet.number_of_mnemonics * 4 / 3);
+        calculate_wallet_id(wallet_id, mnemo);
+        status = memcmp(wallet.wallet_id, wallet_id, WALLET_ID_SIZE) == 0 ? 1 : 0;
```
Comment (amanCypherock) [inline src/level_four/tap_cards/controller/card_action_controllers.c:154]: Log wallet id mismatch result
```suggestion
        status = memcmp(wallet.wallet_id, wallet_id, WALLET_ID_SIZE);
        LOG_INFO("xxx36: %d", status);
        status = (status == 0) ? 1 : 0;
```

File: src/level_four/tap_cards/controller/card_action_controllers.c [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+            wallet_shamir_data.mnemonic_shares,
+            wallet_shamir_data.share_x_coords,
+            secret);
+        memzero(wallet_shamir_data.mnemonic_shares, sizeof(wallet_shamir_data.mnemonic_shares));
+    }
+
+    if (status == 1) {
+        // verify wallet id only if secret successfully regenerated
+        mnemonic_clear();
+        const char* mnemo = mnemonic_from_data(wallet.wallet_share_with_mac_and_nonce, wallet.number_of_mnemonics * 4 / 3);
```
Comment (amanCypherock) [inline src/level_four/tap_cards/controller/card_action_controllers.c:152]: Wrong argument used.
mnemo ASSERT check should also  be added

```suggestion
        const char* mnemo = mnemonic_from_data(secret, wallet.number_of_mnemonics * 4 / 3);
```

File: src/level_four/tap_cards/controller/card_action_controllers.c [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+        secret_size = BLOCK_SIZE;
+        secret = (uint8_t *) malloc(BLOCK_SIZE * sizeof(uint8_t));
+        ASSERT(secret != NULL);
+        status = generate_shares_5C2(
+            wallet_shamir_data.mnemonic_shares,
+            wallet_shamir_data.share_x_coords,
+            secret);
+        memzero(wallet_shamir_data.mnemonic_shares, sizeof(wallet_shamir_data.mnemonic_shares));
+    }
+
```
Comment (amanCypherock) [inline src/level_four/tap_cards/controller/card_action_controllers.c:148]: Log result for 5C2 verification

**Review by amanCypherock — CHANGES_REQUESTED** [review]
Wallet creation failed due to

File: src/level_four/tap_cards/controller/card_action_controllers.c (hunk 3 above)
Comment (amanCypherock) [reply src/level_four/tap_cards/controller/card_action_controllers.c:152]: Also need to test if the variable `wallet.number_of_mnemonics` is set correctly.

File: src/level_four/tap_cards/controller/card_action_controllers.c (hunk 4 above)
Comment (ujjwal-cyph) [reply src/level_four/tap_cards/controller/card_action_controllers.c:148]: 5C2 failure is already logged. should i add here too?

File: src/level_four/tap_cards/controller/card_action_controllers.c (hunk 3 above)
Comment (ujjwal-cyph) [reply src/level_four/tap_cards/controller/card_action_controllers.c:152]: its fetched from cards everytime. `validate_wallet` checks that already. should i add here too?

Comment (amanCypherock) [reply src/level_four/tap_cards/controller/card_action_controllers.c:152]: No check is not required in that case, just add the incorrect variable fix

File: src/level_four/tap_cards/controller/card_action_controllers.c (hunk 4 above)
Comment (amanCypherock) [reply src/level_four/tap_cards/controller/card_action_controllers.c:148]: Not required

File: src/level_four/tap_cards/controller/write_to_cards_controller.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (amanCypherock) [reply src/level_four/tap_cards/controller/write_to_cards_controller.c:93]: Ok, I guess the enums already make it readable enough to not require core logic update for now

**Review by amanCypherock — CHANGES_REQUESTED** [review]
Need to merge develop into this branch

**Comment by ujjwal-cyph** [comment]
> Need to merge develop into this branch and update base to merge into develop

changed

**Review by amanCypherock — CHANGES_REQUESTED** [review]
Need to add INVALID Wallet check on bootup.

File: src/level_four/tap_cards/controller/card_action_controllers.c [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+    if (status == 1) {
+        // verify wallet id only if secret successfully regenerated
+        mnemonic_clear();
+        const char* mnemo = mnemonic_from_data(secret, wallet.number_of_mnemonics * 4 / 3);
```
Comment (amanCypherock) [inline src/level_four/tap_cards/controller/card_action_controllers.c:152]:
```suggestion
        const char* mnemo = mnemonic_from_data(secret, wallet.number_of_mnemonics * 4 / 3);
        ASSERT(mnemo != NULL);
```

File: src/level_four/tap_cards/controller/card_action_controllers.c [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+#include "ui_instruction.h"
+#include "ui_message.h"
+#include "wallet_utilities.h"
+#include "wallet.h"
+
+static void handle_retrieve_wallet_success(uint8_t xcor);
+
+void readback_share_from_card(uint8_t xcor)
+{
+    tap_card_data.retries = 5;
```
Comment (amanCypherock) [inline src/level_four/tap_cards/controller/card_action_controllers.c:72]: Should check card presence here to avoid inconsistent message

**Comment by ujjwal-cyph** [comment]
> Need to add INVALID Wallet check on bootup.

added

File: src/level_three/add_wallet/controller/controller_restore_wallet.c [hunk 7]
```diff
@@ -287,20 +288,23 @@ void restore_wallet_controller()
...
         memzero(wallet.wallet_share_with_mac_and_nonce, sizeof(wallet.wallet_share_with_mac_and_nonce));
         memzero(wallet.arbitrary_data_share, sizeof(wallet.arbitrary_data_share));
         memzero(wallet.checksum, sizeof(wallet.checksum));
         memzero(wallet.key, sizeof(wallet.key));
         memzero(wallet.beneficiary_key, sizeof(wallet.beneficiary_key));
         memzero(wallet.iv_for_beneficiary_key, sizeof(wallet.iv_for_beneficiary_key));
-        flow_level.level_one = LEVEL_TWO_OLD_WALLET;
-        flow_level.level_two = LEVEL_THREE_VERIFY_WALLET;
-        flow_level.level_three = 1;
+        reset_flow_level();
```
[the remark this answers is not in the document]
Comment (ujjwal-cyph) [reply src/level_three/add_wallet/controller/controller_restore_wallet.c:306]: added

File: src/level_four/tap_cards/controller/card_action_controllers.c (hunk 6 above)
Comment (ujjwal-cyph) [reply src/level_four/tap_cards/controller/card_action_controllers.c:72]: i do not understand this. what msg do you see which should be prevented?

Comment (amanCypherock) [reply src/level_four/tap_cards/controller/card_action_controllers.c:72]: If you remove the card in between "write to card" and "read from card", the screen "card detected, processing..." stays on the screen which is incosistent

**Review by amanCypherock — APPROVED** [review]

File: src/level_four/tap_cards/controller/card_action_controllers.c (hunk 6 above)
Comment (ujjwal-cyph) [reply src/level_four/tap_cards/controller/card_action_controllers.c:72]: added

**Pull request merged**